### PR TITLE
[SPARK-8203] [SPARK-8204] [SQL] conditional function: least/greatest

### PR DESCRIPTION
**Pull request by adrian-wang**
@chenghao-intel @zhichao-li @qiansl127

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ConditionalExpressionSuite.scala [hunk 1]
```diff
@@ -134,4 +134,21 @@ class ConditionalExpressionSuite extends SparkFunSuite with ExpressionEvalHelper
     checkEvaluation(CaseKeyWhen(literalNull, Seq(c2, c5, c1, c6)), "c", row)
   }
 
+  test("greatest/least") {
+    val row = create_row(1, 2, "a", "b", "c")
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ConditionalExpressionSuite.scala]: we need to test this for all data types.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 2]
```diff
@@ -312,3 +312,63 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
...
+        s"differing types in Least (${children.map(_.dataType)}).")
+    } else {
+      TypeCheckResult.TypeCheckSuccess
+    }
+  }
+
+  override def dataType: DataType = children.head.dataType
+
+  override def eval(input: InternalRow): Any = {
+    val cmp = GreaterThan
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: Can we make it as the property of the class? We don't want to create new instance for every row, right? And don't forget to mark it as `transient`.

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: Oh, sorry, I just notice it's impossible, but after checking the code in `GreateThan`, we can borrow the `ordering = TypeUtils.getOrdering(left.dataType)`

**Comment by chenghao-intel** [comment]
Don't forget to implement the `def foldable`.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 3]
```diff
@@ -1073,11 +1073,30 @@ object functions {
...
+   *
+   * @group normal_funcs
+   * @since 1.5.0
+   */
+  @scala.annotation.varargs
+  def greatest(columnName: String, columnNames: String*): Column =
+    greatest((columnName +: columnNames).map(Column.apply): _*)
+
+  /**
+    * Computes hex value of the given column
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Nit:
/**
 * Computes hex value for the given column.
 *

**Comment by chenghao-intel** [comment]
Can you rebase the code?

**Comment by adrian-wang** [comment]
retest this please.

**Comment by adrian-wang** [comment]
retest this please.

**Comment by adrian-wang** [comment]
@rxin @davies

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 4]
```diff
@@ -1073,15 +1073,34 @@ object functions {
...
   /**
-   * Computes hex value of the given column
+   * Returns the greatest value of the list of values.
    *
-   * @group math_funcs
+   * @group normal_funcs
    * @since 1.5.0
    */
+  @scala.annotation.varargs
+  def greatest(exprs: Column*): Column = Greatest(exprs.map(_.expr): _*)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Should we check that `exprs` have at least two columns? Also for others, it's good to fail fast.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 5]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
...
+        ctx.genComp(dataType, evalChildren(i).primitive, ev.primitive)} > 0)) {
+          ${ev.isNull} = ${evalChildren(i).isNull};
+          ${ev.primitive} = ${evalChildren(i).primitive};
+        }
+      """
+    s"""
+      ${evalChildren.map(_.code).mkString("\n")}
+      boolean ${ev.isNull} = true;
+      ${ctx.javaType(dataType)} ${ev.primitive} = ${ctx.defaultValue(dataType)};
+      ${(0 to children.length - 1).map(updateEval).mkString("\n")}
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: `0 until children.length`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 6]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
...
+        r
+      }
+    })
+  }
+
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
+    val evalChildren = children.map(_.gen(ctx))
+    def updateEval(i: Int): String =
+      s"""
+        if (${ev.isNull} || (!${evalChildren(i).isNull} && ${
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: Could you move `${` to next line?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 7]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
...
+        ctx.genComp(dataType, evalChildren(i).primitive, ev.primitive)} > 0)) {
+          ${ev.isNull} = ${evalChildren(i).isNull};
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: `${ev.isNull} = false;`

Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: if ev.isNull, this will be executed.

**Comment by adrian-wang** [comment]
cc @davies

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 8]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
     }.mkString
   }
 }
+
+case class Least(children: Expression*) extends Expression {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: note: this should be Seq[Expression], not vararg

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 9]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
...
+  require(children.length > 1, "LEAST requires at least 2 arguments, got " + children.length)
+
+  override def nullable: Boolean = children.forall(_.nullable)
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: shouldn't least be nullable if any is nullable?

Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: hive 1.2.0 only return null when all arguments are null.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: If any child is not nullable, the result will not be nullable.

**Comment by adrian-wang** [comment]
oh, it seems this pr has been merged but stay unclosed...
@davies @rxin

**Comment by srowen** [comment]
@adrian-wang probably a sync problem. It was evidently merged into master. You can just close it too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 10]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
     }.mkString
   }
 }
+
+case class Least(children: Seq[Expression]) extends Expression {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: can you add scaladoc explaining the semantics of this expression? also do it for greatest.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 11]
```diff
@@ -1073,15 +1073,41 @@ object functions {
   def floor(columnName: String): Column = floor(Column(columnName))
 
   /**
-   * Computes hex value of the given column
+   * Returns the greatest value of the list of values.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: should also document semantics regarding null values

**Comment by rxin** [comment]
@adrian-wang I found that happening quite a bit lately. Can you close the pr and just submit followup patches to address my comments? thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala [hunk 12]
```diff
@@ -312,3 +313,103 @@ case class CaseKeyWhen(key: Expression, branches: Seq[Expression]) extends CaseW
...
+case class Greatest(children: Seq[Expression]) extends Expression {
+  require(children.length > 1, "GREATEST requires at least 2 arguments, got " + children.length)
+
+  override def nullable: Boolean = children.forall(_.nullable)
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  private lazy val ordering = TypeUtils.getOrdering(dataType)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.map(_.dataType).distinct.count(_ != NullType) > 1) {
```
Comment (yjshen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: Why count against `NullType`? any reason here?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionals.scala]: i think the problem is that null literals are not converted into any non-null type when it happens in greatest.

**Pull request closed without merging**